### PR TITLE
Add UDF support

### PR DESCRIPTION
**Pull request by nonpool**
This pr is a complement to #67 

## solution
After many attempts, I gave up on hack `RowEncoder`, because it always has inexplicable bugs. 
When register UDF, I directly convert `DataTypeWithClass` to spark DataType, and this solution seems to be working well.

## problem
Please see `UDFRegisterTest.kt` line134, when use a complex type as the return value of UDF, we will got an exception(same as line134 link). This exception also exists in java(i test in my local on spark 3.1.1).So this bug we needs to wait for the official spark to fix or we try to hack related classes to fix it(maybe more effort)

## Possible improvements
- Add more unit test. Actually, I don’t know some common cases of UDF. you can give some help for this

### CODE REVIEW

**Comment by nonpool** [comment]
@asm0dey @FelixEngl

**Comment by asm0dey** [comment]
Looks like returning complex classes works in Scala: https://stackoverflow.com/a/59293056 

Are you sure it may not be supported in Kotlin?

File: kotlin-spark-api/2.4/src/main/kotlin/org/jetbrains/kotlinx/spark/api/UDFRegister.kt [hunk 1]
```diff
@@ -0,0 +1,1356 @@
...
+        return functions.callUDF(udfName)
+    }
+}
+
+/**
+ * Registers the [func] with its [name] in [this]
+ */
+@OptIn(ExperimentalStdlibApi::class)
+inline fun <reified R> UDFRegistration.register(name: String, noinline func: () -> R): UDFWrapper0 {
+    register(name, UDF0(func), schema(typeOf<R>()).unWrapper())
```
Comment (asm0dey) [inline kotlin-spark-api/2.4/src/main/kotlin/org/jetbrains/kotlinx/spark/api/UDFRegister.kt:65]: It looks like it makes sense to work not with our generated schema, but with schema from an encoder, isn't it?

Comment (nonpool) [reply kotlin-spark-api/2.4/src/main/kotlin/org/jetbrains/kotlinx/spark/api/UDFRegister.kt:65]: Yes. Actually, it is better than hack `RowEncoder` to adapter our generated schema when register a UDF. reason: 1. `RowEncoder` only need a correctly `DataType`, not our generated schema other info. 2. some  spark action (eg: `.as<Int>()` action) to use UDF result is not though `RowEncoder`,if still use hack class way, we will hack many class and there is no guarantee that hacked classes will work well.

**Comment by nonpool** [comment]
> Why don't you extract the schema from the generated encoder?

I will remove this to other file

**Comment by nonpool** [comment]
> Looks like returning complex classes works in Scala: https://stackoverflow.com/a/59293056
> 
> Are you sure it may not be supported in Kotlin?

I think this is a big plan. If we don’t have the java implementation as a reference, this future should hack many class. i suggest separate this future to other PR

**Comment by asm0dey** [comment]
> I will remove this to other file

no-no, I meant that we have `encoder` method, which allows you to use the schema, which was already processed by spark.

**Comment by asm0dey** [comment]
> I think this is a big plan. If we don’t have the java implementation as a reference, this future should hack many class. i suggest separate this future to other PR

But if it works for the Scala API should we really provide something so limited? I mean people will just use Scala API because they will get weird exceptions in runtime which sounds as extremely bad UX. Like "we support UDFs, but only for couple of classes". And BTW do we throw exception if provided class is just regular data class or Java bean class? Or they will work?

**Comment by nonpool** [comment]
> > I think this is a big plan. If we don’t have the java implementation as a reference, this future should hack many class. i suggest separate this future to other PR
> 
> But if it works for the Scala API should we really provide something so limited? I mean people will just use Scala API because they will get weird exceptions in runtime which sounds as extremely bad UX. Like "we support UDFs, but only for couple of classes". And BTW do we throw exception if provided class is just regular data class or Java bean class? Or they will work?

I created a simple example for this in my repo [spark-udf-explore](https://github.com/nonpool/spark-udf-explore), and got some conclusion:
- UDF return type is not support java bean in java spark or scala spark
- UDF return type is not support normal scala class
- UDF return type is only support case class in scala spark

Nevertheless, I still want to try my best to let users have a complete API. I will try to hack related classes to support return data class in kotlin API

**Comment by nonpool** [comment]
> > I will remove this to other file
> 
> no-no, I meant that we have `encoder` method, which allows you to use the schema, which was already processed by spark.

But our `encoder` result is wrapper by `KDataTypeWrapper` `KComplexTypeWrapper` and so on,  and these wrapper class is not support by `RowEncoder` or other spark native class which handles Datatype. If we use  wrapper class register UDF we will hack more class and so on he same reason as I said above

**Comment by asm0dey** [comment]
And that's why we potentially should use not our schema, but schema, which can be obtained from encoder, I think

**Comment by nonpool** [comment]
> And that's why we potentially should use not our schema, but schema, which can be obtained from encoder, I think

What the `schema can be obtain from encoder` means? I think i have used our encoder to  obtained schema, just unwrapper our schema when register UDF.

**Comment by asm0dey** [comment]
You're calling `schema(typeOf<R>())`, but you could call `encoder<R>().schema()`

**Comment by nonpool** [comment]
> You're calling `schema(typeOf<R>())`, but you could call `encoder<R>().schema()`

A.....  Because these codes are copied, and I always thought that this is the `encoder` used here.  ^-^

**Comment by asm0dey** [comment]
Copying code is usually just a bad idea I believe. It increases WTFps dramatically

**Comment by nonpool** [comment]
After testing, Call `encoder<R>().schema()` to register UDF is not a good way. `encoder<R>()` method included two action: 
1. call `schema(typeOf<R>())` to get schema(type is `DataTypeWithClass`)
2. generate `serializer` and `deserializer` from schema and construct the `ExpressionEncoder`

so `encoder<R>().schema()` is call the `schema()` method of the `ExpressionEncoder` which is :
```
  // The schema after converting `T` to a Spark SQL row. This schema is dependent on the given
  // serializer.
  val schema: StructType = StructType(serializer.map { s =>
    StructField(s.name, s.dataType, s.nullable)
  })
```
it always wrapper a `StructField` in `StructType`. if call `encoder<R>().schema()`  for `String` it return `{"type":"struct","fields":[{"name":"value","type":"string","nullable":true,"metadata":{}}]}`, but UDF register only need  `StringType`. Not only string type and primitive type are like this, complex type also not work.

so, i think we should call `schema(typeOf<R>())`. Unless there are other changes in the complex type.

**Comment by asm0dey** [comment]
On the other hand, we could unwrap this struct, right? And this way we could avoid Spark dealing with our own `DataTypeWithClass` instances

**Comment by nonpool** [comment]
> On the other hand, we could unwrap this struct, right? And this way we could avoid Spark dealing with our own `DataTypeWithClass` instances

Spark does not directly use our `DataTypeWithClass` instances. Because i unwrap thought this:
```
fun DataType.unWrapper(): DataType {
    return when (this) {
        is DataTypeWithClass -> DataType.fromJson(dt().json())
        else -> this
    }
}
```

**Comment by asm0dey** [comment]
Should it work with deeply nested structures?

**Comment by nonpool** [comment]
`List<String>` and `List< primitive type>`  also work well. 
The Complex type seems to be right when register UDF.

**Comment by asm0dey** [comment]
Wowzer, let's just suppress "unused" inspections

**Comment by asm0dey** [comment]
Or just ignore it and et me know when you're ready

**Pull request merged**